src/main/java/com/yelp/nrtsearch/server/luceneserver/RegisterFieldsHandler.java

```diff
@@ -111,16 +113,34 @@ public FieldDefResponse handle(final IndexState indexState, FieldDefRequest fiel
           throw new RuntimeException(e);
         }
 
-        pendingFieldDefs.put(
-            fieldName, parseOneFieldType(indexState, pendingFieldDefs, fieldName, currentField));
+        FieldDef fieldDef =
+            parseOneFieldType(indexState, pendingFieldDefs, fieldName, currentField);
+        if (currentField.getKey()) {
+          if (keyableField != null) {
+            throw new RegisterFieldsException(
+                String.format(
+                    "field \"%s\" cannot be registered as as key as \"%s\" is already declared as a key field",
+                    currentField.getName(), keyableField.getName()));
+          }
+          if (!(fieldDef instanceof Keyable)) {
+            throw new RegisterFieldsException(
+                "field \"" + fieldName + "\" is not a keyable field type");
+          }
+          keyableField = (Keyable) fieldDef;
+        }
+        pendingFieldDefs.put(fieldName, fieldDef);
       }
     }
 
     // add FieldDef and its corresponding JsonObject to states variable in IndexState
     for (Map.Entry<String, FieldDef> ent : pendingFieldDefs.entrySet()) {
       JsonObject fieldAsJsonObject =
           jsonParser.parse(saveStates.get(ent.getKey())).getAsJsonObject();
-      indexState.addField(ent.getValue(), fieldAsJsonObject);
+      FieldDef fieldDef = ent.getValue();
+      indexState.addField(fieldDef, fieldAsJsonObject);
+    }
+    if (!indexState.hasKeyableField()) {
+      indexState.setKeyableField(keyableField);
     }
```

Comment:
Should keyableField be checked for null too? It looks like passing null to setKeyableField is a noop, but might be better to avoid the call entirely.